core/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java

```diff
@@ -20,12 +20,44 @@
 package org.elasticsearch.common.logging;
 
 import org.elasticsearch.common.SuppressLoggerChecks;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+
+import java.util.Iterator;
+import java.util.Set;
+import java.util.concurrent.CopyOnWriteArraySet;
 
 /**
  * A logger that logs deprecation notices.
  */
 public class DeprecationLogger {
 
+    /**
+     * The "Warning" Header comes from RFC-7234. As the RFC describes, it's generally used for caching purposes, but it can be
+     * used for <em>any</em> warning.
+     *
+     * https://tools.ietf.org/html/rfc7234#section-5.5
+     */
+    public static final String DEPRECATION_HEADER = "Warning";
+
+    /**
+     * This is set once by the {@code Node} constructor, but it uses {@link CopyOnWriteArraySet} to ensure that
+     * tests can run in parallel.
+     */
+    private static final CopyOnWriteArraySet<ThreadContext> THREAD_CONTEXT = new CopyOnWriteArraySet<>();
```

Comment:
I know why you did this, but I think more details in the comments would help ensure the reasoning is clear. Things like tests spawn multiple nodes with separate injectors but the classes come from the same classloader so static variables get shared between nodes.

I wonder if we can workaround the underlying issue in the test framework by creating each node in its own classloader inside of ```InternalTestCluster#buildNode``` so that statics don't get shared